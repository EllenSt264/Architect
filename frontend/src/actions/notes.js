import axios from "axios";
import { GET_NOTES, GET_NOTE, NOTES_LOADING, NOTE_REQUEST_FAIL } from "./types";

export const getNotes = () => async (dispatch) => {
  dispatch({ type: NOTES_LOADING });
  try {
    const res = await axios.get("/api/notes/");
    dispatch({
      type: GET_NOTES,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: NOTE_REQUEST_FAIL,
      payload: err.message,
    });
  }
};

export const getNote = (id) => async (dispatch) => {
  dispatch({ type: NOTES_LOADING });
  try {
    const res = await axios.get(`/api/notes/${id}/`);
    dispatch({
      type: GET_NOTE,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: NOTE_REQUEST_FAIL,
      payload: err.message,
    });
  }
};
