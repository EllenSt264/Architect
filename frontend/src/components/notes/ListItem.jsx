import React from "react";
import { Link } from "react-router-dom";

const constructTitle = (note) => {
  var title = note.split("\n")[0];
  title = title.length > 45 ? title.slice(0, 45) : title;
  return title.trim();
};

const constructSummary = (note) => {
  var title = constructTitle(note);
  var content = note.replaceAll("\n", " ");
  content = content.replaceAll(title, "");
  content = content.length > 40 ? content.slice(0, 40) : content;
  return content.trim();
};

const constructTime = (note) => {
  return new Date(note).toLocaleDateString();
};

const ListItem = ({ note }) => {
  return (
    <Link
      to={`/note/${note.id}`}
      className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
    >
      <h3 className="noteCard-title font-inter capitalize text-white dark:text-white text-1xl tracking-wide">
        {constructTitle(note.body)}
      </h3>
      <p>
        <span className="noteCard-summary text-gray-300 dark:text-gray-200 tracking-wide">
          {constructSummary(note.body)}
        </span>
        <span className="noteCard-time text-orange-200 dark:text-gray-300 font-roboto text-sm tracking-wide ml-2">
          {constructTime(note.updated)}
        </span>
      </p>
    </Link>
  );
};

export default ListItem;
