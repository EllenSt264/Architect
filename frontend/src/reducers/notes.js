import {
  GET_NOTES,
  GET_NOTE,
  ADD_NOTE,
  EDIT_NOTE,
  NOTES_LOADING,
  NOTE_REQUEST_FAIL,
} from "../actions/types";

const initialState = {
  allnotes: [],
  targetNote: {},
  loading: false,
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case NOTES_LOADING:
      return {
        ...state,
        loading: true,
        error: null,
      };
    case GET_NOTES:
      const notesArr = [];
      action.payload.forEach((element) => {
        notesArr.push(element);
      });
      return {
        ...state,
        allnotes: notesArr,
        loading: false,
      };
    case GET_NOTE:
    case ADD_NOTE:
    case EDIT_NOTE:
      return {
        ...state,
        targetNote: action.payload,
        loading: false,
      };
    case NOTE_REQUEST_FAIL:
      return {
        ...state,
        allnotes: [],
        targetNote: {},
        loading: false,
        error: action.payload,
      };
    default:
      return state;
  }
};
