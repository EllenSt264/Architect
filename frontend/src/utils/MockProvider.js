import React from "react";
import PropTypes from "prop-types";
import { MemoryRouter, Switch } from "react-router-dom";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { store } from "../store";

const initialState = {
  notes: {
    allnotes: [],
    targetNote: {},
    loading: false,
    error: null,
  },
};
const mockStore = configureStore([thunk]);

const MockProvider = (props) => {
  const { children, customStore } = props;

  return <Provider store={customStore || store}>{children}</Provider>;
};

MockProvider.propTypes = {
  children: PropTypes.node,
  customStore: PropTypes.shape({}),
};

MockProvider.defaultProps = {
  children: null,
  customStore: mockStore(initialState),
};

export default MockProvider;
