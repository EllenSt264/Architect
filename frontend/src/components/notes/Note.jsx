import React, { Component } from "react";
import { connect } from "react-redux";

import { getNote, addNote, editNote } from "../../actions/notes";

export class Note extends Component {
  constructor(props) {
    super(props);

    const { targetNote } = this.props;
    this.state = {
      targetNote,
      dataLoaded: false,
      isNew: false,
      isChanged: false,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    if (this.props.match.params.id === "new") {
      this.setState({ isNew: true });
    }
    this.loadData();
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  loadData = async () => {
    await this.props.getNote(this.props.match.params.id);
    const { targetNote } = this.props;
    this.setState({
      targetNote: targetNote,
    });
    this.timer = setTimeout(() => {
      this.hasLoaded();
    }, 300);
  };

  hasLoaded() {
    this.setState({ dataLoaded: true });
  }

  handleChange(event) {
    this.setState({
      targetNote: { body: event.target.value },
      isChanged: true,
    });
  }

  handleSubmit = async () => {
    const { id } = this.props.match.params;
    const { targetNote } = this.state;

    if (id === "new" && Object.keys(targetNote).length !== 0) {
      await this.props.addNote(targetNote.body);
    } else if (id !== "new") {
      await this.props.editNote(id, targetNote.body);
    }

    this.props.history.push("/");
  };

  render() {
    return (
      <>
        {!this.state.dataLoaded && !this.state.isNew && (
          <div className="h-screen w-screen pt-8 pl-5 bg-[#8984f6] dark:bg-[#1e163a]">
            Loading...
          </div>
        )}
        {(this.state.dataLoaded || this.state.isNew) && (
          <div className="h-screen w-screen pt-8 pl-5 bg-[#8984f6] dark:bg-[#1e163a]">
            {/* Header */}
            <div className="flex justify-between">
              {/* Back button */}
              <button onClick={this.handleSubmit}>
                <svg
                  className="w-7 h-7 text-white"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>back</title>
                  <path
                    fillRule="evenodd"
                    d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
                    clipRule="evenodd"
                  />
                </svg>
              </button>
              {/* Delete and submit buttons */}
              <div>
                {/* Delete modal */}
                {!this.state.isNew && (
                  <button type="button" className="mx-3 mr-10 text-amber-100">
                    <svg
                      className="w-6 h-6"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <title>delete_modal</title>
                      <path
                        fillRule="evenodd"
                        d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
                        clipRule="evenodd"
                      />
                    </svg>
                  </button>
                )}
                {/* Submit changes */}
                {(this.state.isChanged || this.state.isNew) && (
                  <button
                    className="w-10 text-center ml-4 mr-10 align-top text-amber-200 font-inter font-semibold md:font-bold"
                    onClick={this.handleSubmit}
                  >
                    Done
                  </button>
                )}
              </div>
            </div>
            {/* Textarea */}
            <div className="my-5 flex flex-wrap h-[91%]">
              <textarea
                name="noteBody"
                id="noteBody"
                value={this.state.targetNote?.body}
                onChange={this.handleChange}
                placeholder="Start typing..."
                className="resize-none h-full p-10 flex-1 bg-[#8984f6] dark:bg-[#1e163a] text-gray-100 placeholder:text-gray-300 placeholder:font-inter border-none focus:border-none focus:outline-none font-inter tracking-wide"
              ></textarea>
            </div>
          </div>
        )}
      </>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const noteMatch = Array.isArray(state.notes.allnotes)
    ? state.notes.allnotes.filter((obj) => obj.id == ownProps.match.params.id)
    : "";
  return {
    targetNote: Object.assign({}, ...noteMatch),
  };
};

export default connect(mapStateToProps, { getNote, addNote, editNote })(Note);
