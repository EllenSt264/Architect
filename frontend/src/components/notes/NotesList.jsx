import React, { Component } from "react";
import { connect } from "react-redux";

import { getNotes } from "../../actions/notes";

export class NotesList extends Component {
  componentDidMount() {
    this.props.getNotes();
  }
  render() {
    return (
      <div id="notesList">
        {this.props.notes?.length > 0 &&
          this.props.notes?.map((note) => <p key={note.id}>{note.body}</p>)}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  notes: state.notes.allnotes,
});

export default connect(mapStateToProps, { getNotes })(NotesList);
