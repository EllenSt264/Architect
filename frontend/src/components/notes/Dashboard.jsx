import React, { Component } from "react";

import NotesList from "./NotesList";
import AddButton from "../AddButton";

export class Dashboard extends Component {
  render() {
    return (
      <div
        id="notesContainer"
        className="bg-gradient-to-b dark:bg-transparent from-violet-500 to-cyan-300 dark:bg-gradient-to-tl dark:from-gray-700 dark:via-gray-900 dark:to-black w-full h-full min-h-screen pt-2 pb-8"
      >
        <div className="text-center pt-8 pb-4 font-inter">
          <h1 className="uppercase text-2xl md:text-4xl font-semibold md:font-bold tracking-wider dark:text-white text-yellow-300">
            Architect
          </h1>
          <h2 className="text-lg md:text-2xl md:font-semibold tracking-wide uppercase text-yellow-200 dark:text-gray-200">
            Your life planner
          </h2>
        </div>
        <div className="relative my-5">
          <div className="text-left text-xl font-semibold font-inter md:font-bold tracking-wider dark:text-white text-yellow-200 pl-5">
            <h3>Notes</h3>
          </div>
          <div className="absolute top-0 md:-top-1 right-4 h-6 w-6">
            <AddButton />
          </div>
        </div>
        <div>
          <NotesList />
        </div>
      </div>
    );
  }
}

export default Dashboard;
