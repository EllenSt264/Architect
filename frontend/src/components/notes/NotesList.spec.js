import React from "react";
import { MemoryRouter } from "react-router-dom";

import { mount } from "enzyme";
import toJson from "enzyme-to-json";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import MockProvider from "../../utils/MockProvider";
import NotesList from "./NotesList";
import ListItem from "./ListItem";
import notes from "../../reducers/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("NotesList", () => {
  let wrapper;

  const myMockStore = {
    notes: {
      allnotes: [
        {
          id: 1,
          body: "foo",
          updated: "2022-05-16T07:00:00",
        },
        {
          id: 2,
          body: "Hello react! \n\nHere's some text. ",
          updated: "2022-05-18T10:45:00",
        },
      ],
    },
  };

  beforeEach(() => {
    wrapper = mount(
      <MemoryRouter initialEntries={["/"]}>
        <NotesList />
      </MemoryRouter>,
      { wrappingComponent: MockProvider }
    );
    const provider = wrapper.getWrappingComponent();
    provider.setProps({ customStore: mockStore(myMockStore) });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render with the given state from Redux store", () => {
    // make a snapshot from the component itself, not the router
    const component = wrapper.find(NotesList);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should update the component's props to match the store", () => {
    expect(wrapper.children().props().allnotes).toEqual(myMockStore.allnotes);
  });

  it("should display a list of all notes from store", () => {
    const notesContainer = wrapper.find("#notesList");
    const noteTitleDiv = notesContainer.find(".noteCard-title");
    const noteSummaryDiv = notesContainer.find(".noteCard-summary");
    const noteTimeDiv = notesContainer.find(".noteCard-time");

    // assertions for first note
    expect(noteTitleDiv.at(0).text()).toEqual("foo");
    expect(noteSummaryDiv.at(0).text()).toEqual("");
    expect(noteTimeDiv.at(0).text()).toEqual("16/05/2022");

    // assertions for second note
    expect(noteTitleDiv.at(1).text()).toEqual("Hello react!");
    expect(noteSummaryDiv.at(1).text()).toEqual("Here's some text.");
    expect(noteTimeDiv.at(1).text()).toEqual("18/05/2022");
  });

  it("should render the ListItem component", () => {
    expect(wrapper.containsMatchingElement(<ListItem />)).toEqual(true);
  });
});
