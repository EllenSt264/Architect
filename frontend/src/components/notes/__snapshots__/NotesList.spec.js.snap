// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesList should render with the given state from Redux store 1`] = `
<Connect(NotesList)>
  <NotesList
    getNotes={[Function]}
    notes={
      Array [
        Object {
          "body": "foo",
          "id": 1,
          "updated": "2022-05-16T07:00:00",
        },
        Object {
          "body": "Hello react! 

Here's some text. ",
          "id": 2,
          "updated": "2022-05-18T10:45:00",
        },
      ]
    }
  >
    <div
      id="notesList"
    >
      <ListItem
        key="1"
        note={
          Object {
            "body": "foo",
            "id": 1,
            "updated": "2022-05-16T07:00:00",
          }
        }
      >
        <Link
          className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
          to="/note/1"
        >
          <LinkAnchor
            className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
            href="/note/1"
            navigate={[Function]}
          >
            <a
              className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
              href="/note/1"
              onClick={[Function]}
            >
              <h3
                className="noteCard-title font-inter capitalize text-white dark:text-white text-1xl tracking-wide"
              >
                foo
              </h3>
              <p>
                <span
                  className="noteCard-summary text-gray-300 dark:text-gray-200 tracking-wide"
                />
                <span
                  className="noteCard-time text-orange-200 dark:text-gray-300 font-roboto text-sm tracking-wide ml-2"
                >
                  16/05/2022
                </span>
              </p>
            </a>
          </LinkAnchor>
        </Link>
      </ListItem>
      <ListItem
        key="2"
        note={
          Object {
            "body": "Hello react! 

Here's some text. ",
            "id": 2,
            "updated": "2022-05-18T10:45:00",
          }
        }
      >
        <Link
          className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
          to="/note/2"
        >
          <LinkAnchor
            className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
            href="/note/2"
            navigate={[Function]}
          >
            <a
              className="block p-4 mx-3 my-4 box-border shadow-md dark:shadow-violet-600/25 rounded-md bg-gradient-to-r from-[#8352f1] to-[#95ddff] dark:bg-transparent dark:from-gray-900 dark:via-violet-900 dark:to-gray-900"
              href="/note/2"
              onClick={[Function]}
            >
              <h3
                className="noteCard-title font-inter capitalize text-white dark:text-white text-1xl tracking-wide"
              >
                Hello react!
              </h3>
              <p>
                <span
                  className="noteCard-summary text-gray-300 dark:text-gray-200 tracking-wide"
                >
                  Here's some text.
                </span>
                <span
                  className="noteCard-time text-orange-200 dark:text-gray-300 font-roboto text-sm tracking-wide ml-2"
                >
                  18/05/2022
                </span>
              </p>
            </a>
          </LinkAnchor>
        </Link>
      </ListItem>
    </div>
  </NotesList>
</Connect(NotesList)>
`;
